doc/dev/sharp_bits.rst

```diff
@@ -0,0 +1,1015 @@
+    on **static variables**, those whose values are known.
+
+.. note::
+
+    A general guideline when working with JIT compilation and Catalyst:
+
+    - Python control flow and third party libraries like NumPy and SciPy will
+      be evaluated at compile-time, and can only accept static variables.
+
+    - JAX functions, such as ``jax.numpy``, and Catalyst functions like
+      :func:`~.cond` and :func:`~.for_loop` will be evaluated at
+      runtime, and can accept dynamic variables.
+
+    Note that if AutoGraph is enabled, Catalyst will attempt to convert Python
+    control flow to its Catalyst equivalent to support dynamic variables.
+
+For example, consider the following:
+
+>>> @qjit
+... def f(x):
+...     if x > 5:
+...       x = x / 2
+...     return x ** 2
+>>> f(2.)
+TracerBoolConversionError: Attempted boolean conversion of traced array with shape bool[]..
+The error occurred while tracing the function f at <ipython-input-15-2aa7bf60efbb>:1 for make_jaxpr.
+This concrete value was not available in Python because it depends on the value of the argument x.
+See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerBoolConversionError
+
+This function will fail, as the Python ``if`` statement cannot accept a
+dynamic variable (a JAX tracer) as an argument.
+
+Instead, we can use Catalyst control flow :func:`~.cond` here:
+
+>>> @qjit
+... def f(x):
+...     @cond(x > 5.)
+...     def g():
+...         return x / 2
+...     @g.otherwise
+...     def h():
+...         return x
+...     return g() ** 2
+>>> f(2.)
+array(4.)
+>>> f(6.)
+array(9.)
+
+Here, both conditional branches are compiled, and only evaluated at runtime
+when the value of ``x`` is known.
+
+.. note::
+
+    AutoGraph is an experimental feature that converts Python control flow
+    that depends on dynamic variables to Catalyst control flow behind the
+    scenes:
+
+
+    >>> @qjit(autograph=True)
+    ... def f(x):
+    ...     if x > 5.:
+    ...         print(x)
+    ...         x = x / 2
+    ...     return x ** 2
+    >>> Traced<ShapedArray(float64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>
+    ... array(4.)
+    >>> f(6.)
+    ... array(9.)
+
+    For more details, see the AutoGraph guide.
+
+Note that, if the Python ``if`` statement depends only on values that are
+static (known at compile time), this is fine --- the ``if`` statement will
+simply be evaluated at compile time rather than runtime:
+
+>>> @qjit
+... def f(x):
+...     for i in range(2):
+...         print(i, x)
+...         x = x / 2
+...     return x ** 2
+>>> f(2.)
+0 Traced<ShapedArray(float64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>
+1 Traced<ShapedArray(float64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>
+array(0.25)
+
+Here, the for loop is evaluated at compile time (notice the multiple tracers
+that have been printed out during program capture --- one for each loop!),
+rather than runtime.
```

Comment:
@dime10 yes I wrote this explicitly ignoring Autograph - what changes do you suggest here?

Reply:
For me, I think the main valuable thing was knowing that `for ...:` etc. _may_ still work without autograph, but the behaviour may not be what you intend

Reply:
Oh I see, I think because the preceding note was talking about Autograph I thought this was still in the same context.

Maybe we could move the autograph note to after this paragraph?